### PR TITLE
Add dai pool functions

### PR DESCRIPTION
**Pull request by salazarguille**
It contains:

- DAIPool contract functions (except withdrawInterest).
- Unit tests for all implemented functions.
- Some configuration.

image

image

### CODE REVIEW

File: contracts/base/DAIPool.sol [hunk 1]
```diff
@@ -43,31 +45,49 @@ contract DAIPool is DAIPoolInterface, Initializable {
...
 
     /** External Functions */
 
+    /**
+        @notice It initializes the contract state variables.
+        @param zdaiAddress zDAI token address.
+        @param daiAddress DAI token address.
+        @param lenderInfoAddress LenderInfo contract address.
+        @param loansAddress Loans contract address.
+        @dev It throws a require error if the contract is already iintialized.
```
Comment (alicevhenshaw) [inline contracts/base/DAIPool.sol:72]: typo: `iintialized`

File: contracts/base/DAIPool.sol [hunk 2]
```diff
@@ -89,20 +109,130 @@ contract DAIPool is DAIPoolInterface, Initializable {
...
-    function withdrawDai(uint256 amount) external {}
+    /**
+        @notice It allows any zDAI holder to burn their zDAI tokens and withdraw their DAIs.
+        @param amount of DAI tokens to withdraw.
+        @dev It throws a require error if the contract hasn't enough DAI balance.
+        @dev It throws a require error if the holder hasn't enough zDAI balance.
+     */
+    function withdrawDai(uint256 amount) external {
+        // Checks contract hast enough DAI balance.
+        requireEnoughDAIBalance(amount);
```
Comment (alicevhenshaw) [inline contracts/base/DAIPool.sol:120]: DAI balance check is not needed, as all ERC20 tokens revert on `transfer` if not enough balance is provided.

File: contracts/base/DAIPool.sol [hunk 3]
```diff
@@ -89,20 +109,130 @@ contract DAIPool is DAIPoolInterface, Initializable {
...
+
+        // Check holder/sender has enough zDAI balance.
+        requireEnoughZDaiBalanceFor(msg.sender, amount);
```
Comment (alicevhenshaw) [inline contracts/base/DAIPool.sol:123]: ZDAI balance check is not needed, as _burn reverts if the user does not have enough ZDAI

File: contracts/base/DAIPool.sol [hunk 4]
```diff
@@ -89,20 +109,130 @@ contract DAIPool is DAIPoolInterface, Initializable {
...
+        zdai.burn(amount);
+
+        // Notify ZDAI tokens were burnt/
+        lenderInfo.zDaiBurnt(msg.sender, amount);
+
+        // Transfers DAI tokens
+        daiTransfer(msg.sender, amount);
+
+        // Emit event.
+        emit DaiWithdrew(msg.sender, amount);
```
Comment (alicevhenshaw) [inline contracts/base/DAIPool.sol:135]: The wording `DaiWithdrawn` makes more sense for this scenario.

File: contracts/base/DAIPool.sol [hunk 5]
```diff
@@ -89,20 +109,130 @@ contract DAIPool is DAIPoolInterface, Initializable {
...
+        @param amount in DAI tokens to liquidate.
+        @param liquidator address to receive the tokens.
+        @dev It throws a require error if this contract hasn't enough DAI balance.
+     */
+    function liquidationPayment(uint256 amount, address liquidator)
+        external
+        isLoan(msg.sender)
+    {
+        // Checks contract has enough DAI balance.
+        requireEnoughDAIBalance(amount);
```
Comment (alicevhenshaw) [inline contracts/base/DAIPool.sol:165]: Again this function shouldnt be needed

File: contracts/base/DAIPool.sol [hunk 6]
```diff
@@ -89,20 +109,130 @@ contract DAIPool is DAIPoolInterface, Initializable {
...
+
+        // Transfers DAIs to the liquidator.
+        daiTransfer(liquidator, amount);
+
+        // Emits event
+        emit PaymentLiquidated(liquidator, amount);
+    }
+
+    /**
+        @notice Ønce the loan is created, it transfers the amount of DAIs to the borrower.
```
Comment (alicevhenshaw) [inline contracts/base/DAIPool.sol:175]: Strange symbol by accident I think `Ønce`

File: contracts/base/DAIPool.sol [hunk 7]
```diff
@@ -89,20 +109,93 @@ contract DAIPool is DAIPoolInterface, Initializable {
...
+    function createLoan(uint256 amount, address borrower) external isLoan(msg.sender) {
+        // Transfer DAIs to the borrower.
+        daiTransfer(borrower, amount);
+    }
+
+    /**
+        @notice It gets the current DAI balance for this contract.
+        @return current DAI balance.
+     */
+    function getDaiBalance() external view returns (uint256) {
```
Comment (alicevhenshaw) [inline contracts/base/DAIPool.sol:178]: This function isn't used anywhere in the contract. Is there a reason it's being kept? Might be worth deleting it now.

File: contracts/base/DAIPool.sol [hunk 8]
```diff
@@ -89,20 +109,93 @@ contract DAIPool is DAIPoolInterface, Initializable {
...
+    }
+
+    /**
+        @notice Once the loan is created, it transfers the amount of DAIs to the borrower.
+
+        @param amount of DAI tokens to transfer.
+        @param borrower address which will receive the DAI tokens.
+        @dev This function only can be invoked by the LoansInterface implementation.
+     */
+    function createLoan(uint256 amount, address borrower) external isLoan(msg.sender) {
```
Comment (alicevhenshaw) [inline contracts/base/DAIPool.sol:169]: Should the `isInitialized` identifier be used on more functions? I think it's currently just on `depositDai`

File: contracts/base/DAIPool.sol [hunk 9]
```diff
@@ -43,31 +45,49 @@ contract DAIPool is DAIPoolInterface, Initializable {
...
+    address public loans;
 
     /** Modifiers */
 
+    /**
+        @notice It checks the address is the Loans contract address.
+        @param anAddress to verify.
+        @dev It throws a require error if parameter is not equal to loans contract address.
+     */
+    modifier isLoan(address anAddress) {
```
Comment (alicevhenshaw) [inline contracts/base/DAIPool.sol:57]: This modifier is only ever used with the parameter `msg.sender`. The modifier parameter can therefore be removed, and just be` isLoan()` with `loans.requireEqualTo(msg.sender)`. Much like how `onlyOwner` doesn't take a parameter, it just uses msg.sender.

File: contracts/base/DAIPool.sol (hunk 8 above)
Comment (salazarguille) [reply contracts/base/DAIPool.sol]: Good catch. I have added it to the external functions.

**Pull request merged**